endpoints/openrtb2/auction.go

```diff
@@ -653,17 +654,40 @@ func mapSChains(req *openrtb_ext.RequestWrapper) error {
 	return nil
 }
 
-func validateAndFillSourceTID(req *openrtb2.BidRequest) error {
+func validateAndFillSourceTID(req *openrtb_ext.RequestWrapper, generateRequestID bool, hasStoredBidRequest bool, isAmp bool) error {
 	if req.Source == nil {
 		req.Source = &openrtb2.Source{}
 	}
-	if req.Source.TID == "" {
-		if rawUUID, err := uuid.NewV4(); err == nil {
-			req.Source.TID = rawUUID.String()
-		} else {
-			return errors.New("req.Source.TID missing in the req and error creating a random UID")
+
+	rawUUID, err := uuid.NewV4()
+	if err != nil {
+		return errors.New("error creating a random UUID for source.tid")
+	}
+
+	if req.Source.TID == "" || req.Source.TID == "{{UUID}}" {
+		req.Source.TID = rawUUID.String()
+	}
+
+	if generateRequestID && (isAmp || hasStoredBidRequest) {
+		req.Source.TID = rawUUID.String()
+	}
+
+	for _, impWrapper := range req.GetImp() {
+		ie, _ := impWrapper.GetImpExt()
```

Comment:
Again we should redo the logic here as I suggested above, so we are not calling `uuid.NewV4()` for every imp of every request that passes through the system before even checking if it is needed.

Reply:
updated